### DISCUSSION TITLE
Unable to find binds on base class when using vertical partitioning

### DISCUSSION THREAD

**Discussion by lukeschissler** [sqlalchemy/sqlalchemy#10725] [Usage Questions] [answered]
Hi all!

I'm attempting to use the vertical partitioning strategy described in the docs [here](https://docs.sqlalchemy.org/en/20/orm/persistence_techniques.html#simple-vertical-partitioning) to dynamically resolve queries to two different databases in one session. I'm on sqlalchemy v2.0.23. Here's the code:
```python
from sqlalchemy import BigInteger, create_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, sessionmaker

class BaseLocal(DeclarativeBase):
    pass

class BaseRemote(DeclarativeBase):
    pass

class LocalA(BaseLocal):
    __tablename__ = "localA"
    id: Mapped[int] = mapped_column(BigInteger, unique=True, primary_key=True)


class LocalB(BaseLocal):
    __tablename__ = "localB"
    id: Mapped[int] = mapped_column(BigInteger, unique=True, primary_key=True)


class RemoteA(BaseRemote):
    __tablename__ = "remoteA"
    id: Mapped[int] = mapped_column(BigInteger, unique=True, primary_key=True)


class RemoteB(BaseRemote):
    __tablename__ = "remoteB"
    id: Mapped[int] = mapped_column(BigInteger, unique=True, primary_key=True)


local_engine = create_engine("database_1")
remote_engine = create_engine("database_2")

Session = sessionmaker()
Session.configure(binds={BaseLocal: local_engine, BaseRemote: remote_engine})
session = Session()
```

This works fine when performing queries on specific tables, such as:
```python
session.query(LocalA).all()
```
The query gets correctly routed to database 1 in this instance and everything works as expected.

However, in another use case, I'd like to iterate over all the tables associated with a base's MetaData, such as:
```python
for table in BaseLocal.metadata.tables.values():
    session.query(table).limit(1).first()
```

When I attempt to use that code, I get an error with this traceback:
```bash
Traceback (most recent call last):
  File "/app/hazelnut/scratch.py", line 47, in <module>
    session.query(table).limit(1).first()
  File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2748, in first
    return self.limit(1)._iter().first()  # type: ignore
  File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2847, in _iter
    result: Union[ScalarResult[_T], Result[_T]] = self.session.execute(
  File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2308, in execute
    return self._execute_internal(
  File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2178, in _execute_internal
    bind = self.get_bind(**bind_arguments)
  File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2792, in get_bind
    raise sa_exc.UnboundExecutionError(
sqlalchemy.exc.UnboundExecutionError: Could not locate a bind configured on SQL expression or this Session.
```

I'm only running into this issue with looping over the metadata tables when I use vertical partitioning - if I define a single bind on the `sessionmaker(bind=some_bind)` call, it works as expected.

Am I defining my `Base`'s incorrectly? Is there a better way to iterate over the tables associated with a database when using vertical partitioning?

**Comment by zzzeek** [comment, marked as answer]
hi -

the `session.delete()` method accepts an instance of a mapped class, not a Table object, that operation would raise an error.

The binds you are using are class-based, and without binding to specific classes or tables, such as placing `LocalB` directly in the 'binds' map, the session bind resolution will not work for Table objects like that.

(edited, I see those are two separate bases)

for the use case you have, I would use a delete() SQL statement with each MetaData collection:

```py

with local_engine.begin() as conn:
    for table in reversed(BaseLocal.metadata.sorted_tables):
        conn.execute(table.delete())

with remote_engine.begin() as conn:
    for table in reversed(BaseRemote.metadata.sorted_tables):
        conn.execute(table.delete())

```

**Reply by lukeschissler** [reply]
Hey, thanks for taking a look!

Your solution seemed to work for me - thank you.

I tried it both with a session executing the `table.delete()` - which did not work (same bind error) and as you suggested, with the `local_engine.begin()` context manager, which did work. Is there a reason multiple binds don't play well with sessions?

edit - code snippets

works
```python
with local_engine.begin() as conn:
    for table in reversed(BaseLocal.metadata.sorted_tables):
        conn.execute(table.delete())
``` 

does not work
```python
session = Session()

for table in reversed(BaseLocal.metadata.sorted_tables):
    session.execute(table.delete())
```

**Reply by zzzeek** [reply]
multiple binds play great with sessions but the session is oriented towards using mapped classes, not Table objects, as the lead entities inside of queries, so the lookup for "entity -> superclass is in the binds map" does not take place for a Table object.   it currently would work if you had the direct mapped class in the binds map, so it might be considered a bug that the "superclass" lookup feature does not work for Table that is linked to a class, since this could be made to work now (probably could not work in earlier SQLAlchemy versions).

the reason the session.execute() example you have does not work is for this reason.

**Reply by lukeschissler** [reply]
Ah gotcha, makes perfect sense. 

Is the binds map something I could manipulate manually so that the superclass lookup works?

**Reply by zzzeek** [reply]
you can put all the Table objects into it directly, yes you can update it after the fact like that sure

**Reply by zzzeek** [reply]
actually there's an ~~add_bind()~~ bind_table() method, use that

**Reply by zzzeek** [reply]
[bind_table](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.bind_table)
